crates/bevy_ecs/src/schedule/fsm.rs

```diff
@@ -0,0 +1,113 @@
+use crate::{
+    self as bevy_ecs,
+    change_detection::Mut,
+    schedule::{run_systems, SystemLabel},
+    system::Res,
+    world::World,
+};
+
+use std::{fmt::Debug, hash::Hash};
+use thiserror::Error;
+
+/// Types that can define states in a finite-state machine.
+pub trait State: 'static + Send + Sync + Clone + PartialEq + Eq + Debug + Hash {}
+impl<T> State for T where T: 'static + Send + Sync + Clone + PartialEq + Eq + Debug + Hash {}
+
+/// A [`SystemLabel`] for a state's "on enter" transition.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, SystemLabel)]
+pub struct OnEnter<S: State>(pub S);
+
+/// A [`SystemLabel`] for a state's "on exit" transition.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, SystemLabel)]
+pub struct OnExit<S: State>(pub S);
+
+/// A simple finite-state machine.
```

Comment:
This needs more links explaining context; this is clear in the source, but won't be on docs.rs.